### DISCUSSION TITLE
Clearing STDOUT

### DISCUSSION THREAD

**Discussion by morgverd** [ronf/asyncssh#499] [Q&A]
How would you clear the stdout pipe in a generic SSH server implementation?
Running a forwarded process and executing clear works, so is there a simplistic way of doing that natively?

Thanks.

**Comment by ronf** [comment]
I'm not sure what you mean by "executing clear".

Are you trying to "clear the screen" on the terminal window running the SSH client, where AsyncSSH is acting as an SSH server, or are you trying to discard output from an upstream SSH server instead of outputting it to the terminal with AsyncSSH acting as an SSH client?

**Reply by morgverd** [reply]
Sorry for not being clear. I’m using AsyncSSH as the SSH server and I need to clear the connected SSH client. When I use the IO redirection and pipe to a subprocess and run clear it works, so I’m wondering if there’s a way to do this without having to proxy it to another application.

**Reply by ronf** [reply]
This can be done, but the answer depends on the terminal type of the SSH client.

These days, most terminals are ANSI-compatible, and so it would work to write the ANSI version of the "clear screen" sequence to the SSHServerProcess's stdout. If `process` is your SSHServerProcess object, this would look something like:

```python
    process.stdout.write('\x1b[H\x1b[J')
```

Here, `'\x1b'` is the ESC character, and so the above is basically ESC [ H (move cursor to top left) followed by ESC [ J (clear to end of screen).

This may not work on all terminals, though. If you wanted to be completely portable, you'd have to have code which checked the value of process.term_type and knew what the right sequence was to send for that terminal type. When running a normal application, the "curses" package could be used for this, but since every connection coming into the SSH server could have a different terminal type and "curses" can only support a single terminal, it can't be used in this case. In theory, there might be a way to do your own lookup in the "terminfo" or "termcap" databases to determine what ESC sequences to use for a given terminal, but I'm not sure if there's a portable way to do that in Python.

**Reply by morgverd** [reply]
Thanks the `term_type` is a great addition I didn't even see originally. While I've got you here, I've been experimenting more with the `SSHServerSession` opposed to using a `process_factory`. The main issue I'm running into is simply that the `session_started` doesn't support async, so I can't actually start the tasks like I could with the process factory method. Is there any possible work around here?

OR, Is there an actual way to see when a client disconnects using the `process_factory`?

**Reply by ronf** [reply]
When using process_factory, you are passing in a callable, so there's no way to tell it to use a subclass of SSHServerProcess in this case. However, if your process handler is reading from process.stdin, you'll get an EOF on that as soon as the client closes the session (or the connection). This would show up as the read call returning an empty string.

AsyncSSH may also raise a BrokenPipeError if you try to write to process.stdout or process.stderr on a closed session. Checking for an empty string on read or catching those errors on write calls is the standard way to determine that the session has been closed.

In the case where the whole SSH connection is closed abruptly without cleaning up open sessions first, you may get back a ConnectionLost exception on any outstanding process reads or writes, so you'll want to check for that as well.
